firedrake/petsc.py

```diff
@@ -259,3 +268,79 @@ def inserted_options(self):
         finally:
             for k in self.to_delete:
                 del self.options_object[self.options_prefix + k]
+
+
+def _extract_comm(obj: Any) -> MPI.Comm:
+    """ Extract Firedrake internal comm off given object
+
+    Parameters
+    ----------
+    obj:
+        Any Firedrake object or any comm
+
+    Returns
+    -------
+    MPI.Comm
+        Internal communicator
+
+    """
+    comm = None
+    if isinstance(obj, (PETSc.Comm, mpi.MPI.Comm)):
+        try:
+            if mpi.is_pyop2_comm(obj):
+                comm = obj
+            else:
+                internal_comm = obj.Get_attr(mpi.innercomm_keyval)
+                if internal_comm is None:
+                    comm = obj
+                else:
+                    comm = internal_comm
+        except mpi.PyOP2CommError:
+            pass
+    elif hasattr(obj, "_comm"):
+        comm = obj._comm
+    elif hasattr(obj, "comm"):
```

Comment:
This is not what the docstring says this function does. This isn't the "internal Firedrake comm".

Reply:
Ah, I see the confusion here, I have amended the docstring to say Firedrake/PyOP2 internal comm. This covers the case where the object is actually a "PyOP2 object"